### PR TITLE
Pending

### PR DESCRIPTION
**Pull request by pbrant**

### CODE REVIEW

File: src/Ccap/Codegen/Scala.purs [hunk 1]
```diff
@@ -40,63 +54,104 @@ paren :: Box -> Array Box -> Box
...
         cls = (text "final case class" <<+>> text name) `paren` (recordFieldType <$> props)
-
         enc = defEncoder name (text "x => argonaut.Json.obj" `paren` (recordFieldEncoder <$> props))
-      in cls // enc
-    Sum vs ->
+        decBody =
+          case Array.length props of
+            1 -> maybe (emptyBox 0 0) (singletonRecordDecoder name) (Array.head props)
+            x | x <= 12 -> smallRecordDecoder name props
+            x -> bigRecordDecoder name props
```
Comment (aij) [inline src/Ccap/Codegen/Scala.purs]: If I'm understaning this correctly, this logic should be in `recordDecoder`/`bigRecordDecoder` and evaluated recursively.

The idea is to work around limitations in ScalaZ, right?

File: src/Ccap/Codegen/Scala.purs [hunk 2]
```diff
@@ -127,17 +135,25 @@ typeDecl (TypeDecl name tt an) =
             text "type" <<+>> text name <<+>> char '=' <<+>> text typ
               // wrapEncoder name t (text encode)
               // wrappedDecoder
-    Record props ->
+    Record props -> do
+      Module modName _ _ <- asks _.currentModule
       let
-        cls = (text "final case class" <<+>> text name) `paren` (recordFieldType <$> props)
+        qualify =
+          if modName == name && outputMode == TopLevelCaseClass
```
Comment (aij) [inline src/Ccap/Codegen/Scala.purs:151]: I like destuttering names like this, but I'd prefer if we did it more consistently. The case classes generated from Sum types should be just as easy as records, but type aliases would be trickier since we'd need to compile all relevant files at once...

Would it be easier to just not stutter in the module definitions, so that we don't need to destutter them?

File: src/Ccap/Codegen/Parser.purs [hunk 3]
```diff
@@ -76,13 +76,24 @@ lower = satisfy isLower <?> "lowercase letter"
 identifier :: ParserT String Identity String
 identifier = tokenParser.identifier
 
+-- Quick hack to parse identifiers ignoring reserved words. TODO Figure out what
+-- tokens really need to be reserved (if any).
```
Comment (aij) [inline src/Ccap/Codegen/Parser.purs:80]: Why is this needed? Which identifiers were conflicting with keywords?

**Review by aij — COMMENTED** [review]
A couple quick comments before I go home for the weekend...

dffe67634 looks good, but I still need to read some of the others

Would you mind elaborating a bit more on the reasoning behind these changes?

File: src/Ccap/Codegen/Parser.purs (hunk 3 above)
Comment (pbrant) [reply src/Ccap/Codegen/Parser.purs:80]: I originally ran into it with `module`. `type` is another one that might come up.

File: src/Ccap/Codegen/Scala.purs (hunk 2 above)
Comment (pbrant) [reply src/Ccap/Codegen/Scala.purs:151]: How would that look syntactically? It seems like we meet ourselves coming and going one way or another here since any syntax that's aligns with Scala's way of doing things won't align with PureScript and vice versa.

File: src/Ccap/Codegen/Scala.purs (hunk 1 above)
Comment (pbrant) [reply src/Ccap/Codegen/Scala.purs]: Correct about Scalaz (`applyX` goes up to 12, `tupleX` up to 5).

`bigRecordDecoder` definitely could be improved by continuing to nest tuples as deeply as necessary accompanied with a top level `applyX`. I was thinking 60 fields would be enough for now.
